### PR TITLE
use byte array instead of string for code fetch

### LINKED ISSUE #1113: Pass long strings as [ubyte] on code_fetch
Long strings (e.g. `source_code`, `output_code`) should be passed  as binary `[ubyte]` instead of as a string. Since JS will decode the  flatbuffer into a string anyway, having these be a string means Rust must also decode `utf8`.

`CodeFetchRes` contains some of these:

https://github.com/denoland/deno/blob/a68403d09006af3e7f75aacb9e818ce3bcf60b13/src/msg.fbs#L142-L151

### PR DESCRIPTION
**Pull request by F001**
Fix #1113

cc @kitsonk

### CODE REVIEW

File: src/deno_dir.rs [hunk 1]
```diff
@@ -453,19 +451,26 @@ fn test_code_cache() {
...
+  const DIGIT_TO_CHAR: [char; 16] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e',
+    'f',
+  ];
   for byte in digest.as_ref() {
-    write!(&mut out, "{:02x}", byte).unwrap();
+    let low = (byte & 0x0F) as usize;
+    let high = (byte >> 4) as usize;
+    out.push(DIGIT_TO_CHAR[high]);
+    out.push(DIGIT_TO_CHAR[low]);
```
Comment (ry) [inline src/deno_dir.rs]: I'm not sure this is better?

File: src/msg.fbs [hunk 2]
```diff
@@ -157,18 +157,16 @@ table CodeFetchRes {
...
   filename: string;
   media_type: MediaType;
-  // TODO These should be [ubyte].
-  // See: https://github.com/denoland/deno/issues/1113
-  source_code: string;
-  output_code: string; // Non-empty only if cached.
-  source_map: string; // Non-empty only if cached.
+  source_code: [ubyte];
+  output_code: [ubyte]; // Non-empty only if cached.
+  source_map: [ubyte]; // Non-empty only if cached.
```
Comment (ry) [inline src/msg.fbs]: 👍

File: src/deno_dir.rs (hunk 1 above)
Comment (F001) [reply src/deno_dir.rs]: It is worse in terms of readability. It is a tradeoff.

Comment (ry) [reply src/deno_dir.rs]: What's the tradeoff? Is this faster?

Comment (F001) [reply src/deno_dir.rs]: Yes, definitely faster. Formatter involves very complex execution path.

Comment (ry) [reply src/deno_dir.rs]: I'm skeptical - certainly this is faster - but I doubt it's 2X faster. Anyway it doesn't matter. This is readable enough.

Comment (F001) [reply src/deno_dir.rs]: I bet 10x faster. The default impl of `ToString` trait is based on formatter. It is quite slow. That is why many types has its own version of specialized impl. This was a big motivation of specilization feature.

Comment (kitsonk) [reply src/deno_dir.rs]: wow, interesting...

File: js/os.ts [hunk 3]
```diff
@@ -48,13 +49,16 @@ export function codeFetch(
...
+  const outputCode = codeFetchRes.outputCodeArray() || undefined;
+  const sourceMap = codeFetchRes.sourceMapArray() || undefined;
   return {
     moduleName: codeFetchRes.moduleName() || undefined,
     filename: codeFetchRes.filename() || undefined,
     mediaType: codeFetchRes.mediaType(),
-    sourceCode: codeFetchRes.sourceCode() || undefined,
-    outputCode: codeFetchRes.outputCode() || undefined,
-    sourceMap: codeFetchRes.sourceMap() || undefined
+    sourceCode: sourceCode && new TextDecoder().decode(sourceCode),
```
Comment (kitsonk) [inline js/os.ts]: one decoder up above likely would be more memory efficient... (slightly) instead of newing up a new decoder each property

Comment (F001) [reply js/os.ts]: Good idea!

File: src/deno_dir.rs (hunk 1 above)
Comment (F001) [reply src/deno_dir.rs]: I accidentally find that "Hex encoding" is something very suitable for SIMD, it is used as an example in [standard document](https://doc.rust-lang.org/std/arch/index.html#examples). My version is the `hex_encode_fallback`.

I would leave this optimization for future work.

**Pull request merged**